xhydro/modelling/calibration.py

```diff
@@ -0,0 +1,322 @@
+
+It then becomes the user's responsibility to ensure that required data for a
+given model be provided in the model_config object both in the data preparation
+stage and in the hydrological model implementation. This can be addressed by
+a set of pre-defined codes for given model structures.
+
+For example, for GR4J, only small datasets are required and can be stored
+directly in the model_config dictionary. However, for Hydrotel or Raven models,
+maybe it is better to pass paths to netcdf files which can be passed to the
+models. This will require pre- and post-processing, but this can easily be
+handled at the stage where we create a hydrological model and prepare the data.
+
+The calibration aspect then becomes trivial:
+
+    1. A model_config object is passed to the calibrator.
+    2. Lower and upper bounds for calibration parameters are defined and passed
+    3. An objective function, optimizer and hyperparameters are also passed.
+    4. The calibrator uses this information to develop parameter sets that are
+       then passed as inputs to the "model_config" object.
+    5. The calibrator launches the desired hydrological model with the
+       model_config object (now containing the parameter set) as input.
+    6. The appropriate hydrological model function then parses "model_config",
+       takes the parameters and required data, launches a simulation and
+       returns simulated flow (Qsim).
+    7. The calibration package then compares Qobs and Qsim and computes the
+       objective function value, and returns this to the sampler that will then
+       repeat the process to find optimal parameter sets.
+    8. The code returns the best parameter set, objective function value, and
+       we also return the simulated streamflow on the calibration period for
+       user convenience.
+
+This system has the advantage of being extremely flexible, robust, and
+efficient as all data can be either in-memory or only the reference to the
+required datasets on disk is passed around the callstack.
+
+Currently, the model_config object has 3 mandatory keywords for the package to
+run correctly in all instances:
+
+    - model_config["Qobs"]: Contains the observed streamflow used as the
+                            calibration target.
+
+    - model_config["model_name"]: Contains a string referring to the
+                                  hydrological model to be run.
+
+    - model_config["parameters"]: While not necessary to provide this, it is
+                                  a reserved keyword used by the optimizer.
+
+Any comments are welcome!
+"""
+
+# Import packages
+import numpy as np
+import spotpy
+from spotpy import analyser
+from spotpy.parameter import Uniform
+
+from xhydro.modelling.hydrological_modelling import hydrological_model_selector
+from xhydro.modelling.obj_funcs import (
+    get_objective_function,
+    get_objfun_minimize_or_maximize,
+    get_optimizer_minimize_or_maximize,
+)
+
+
+class spot_setup:
```

Comment:
```suggestion
class Spot_setup:
```
Les classes commencent avec une majuscule.

Reply:
I'd go even further:
```suggestion
class SpotSetup:
```

Reply:
Done!